docs/app/Examples/collections/Form/Types/FormAdvancedExample.js

```diff
@@ -0,0 +1,243 @@
+    const { values } = this.state
+    const { name, schema } = node
+    const { prefix = [] } = options
+
+    const keyPath = [...prefix, name].join('.')
+    const items = _.get(values, keyPath) || []
+
+    const renderItem = (_item, index) => {
+      const itemPrefix = [...prefix, name, index]
+
+      return (
+        <Form.Group widths='equal'>
+          {this.renderFormSchema(schema, { ...options, prefix: itemPrefix })}
+          <Button onClick={this.handleNestedRemove} name={createKeyPath(itemPrefix)} content='Remove' />
+        </Form.Group>
+      )
+    }
+
+    return (
+      <Form.Field>
+        <label>{labelFromNode(node)}</label>
+        {items.map(renderItem)}
+        <Button onClick={this.handleNestedAdd} name={keyPath} content='Add' />
+      </Form.Field>
+    )
+  }
+
+  renderFormGroup (node, options = {}) {
+    const { schema } = node
+
+    return <Form.Group widths='equal'>{this.renderFormSchema(schema, options)}</Form.Group>
+  }
+
+  renderFormNode (node, options = {}) {
+    const { values } = this.state
+    const { name, placeholder, type } = node
+    const { prefix = [] } = options
+
+    if (type === 'group') {
+      return this.renderFormGroup(node, options)
+    }
+
+    if (type === 'nested') {
+      return this.renderFormCollection(node, options)
+    }
+
+    const keyPath = createKeyPath([...prefix, name])
+
+    const baseFieldProps = {
+      control: typeToControl[type],
+      name: keyPath
+    }
+
+    let fieldProps
+
+    if (baseFieldProps.control === Input) {
+      fieldProps = {
+        label: labelFromNode(node),
+        onChange: this.handleInputChange,
+        placeholder: placeholder || _.startCase(name),
+        value: _.get(values, keyPath) || ''
+      }
+    } else if (baseFieldProps.control === Checkbox) {
+      fieldProps = {
+        label: labelFromNode(node),
+        onChange: this.handleCheckboxChange,
+        placeholder: placeholder || _.startCase(name),
+        checked: !!_.get(values, keyPath)
+      }
+    } else if (baseFieldProps.control === Dropdown) {
+      fieldProps = {
+        label: labelFromNode(node),
+        onChange: _.partial(this.handleDropdownChange, name),
+        placeholder: placeholder || _.startCase(name),
+        value: _.get(values, keyPath) || '',
+        options: node.options
+      }
+    } else if (type === 'submit') {
+      fieldProps = {
+        onClick: this.handleSubmit,
+        content: labelFromNode(node)
+      }
+    }
+
+    return <Form.Field {...baseFieldProps} {...fieldProps} />
+  }
+
+  renderFormSchema (schema, options) {
+    return schema.map((node) => this.renderFormNode(node, options))
+  }
+
+  render () {
+    const { formSchema } = this.props
+
+    return (
+      <Form>
+        {this.renderFormSchema(formSchema)}
+        <Message>
+          <pre>state: {JSON.stringify(this.state, null, 2)}</pre>
+        </Message>
+      </Form>
+    )
+  }
+}
+
+const formSchema = [
```

Comment:
I like this, let's iterate on it and see where it goes!  Here's my first proposal:

- each object should be spread on the control
- explicitly define the field `control` prop, instead of reusing `type`
- use an array of `fields` where nested arrays would be groups
- default the `control` to `Input` and `type` to `text`
- capitalized `control` values will use the Stardust control, lower case for HTML controls (e.g. `input` vs `Input`)

This would make it more inline with our other shorthand.  It also gives the user more control, we could allow them to pass an actual `control` component and it's props, just like the Field does.  Finally, we're just spreading prop objects, a good thing.

I've left out the email `nested` part for now.  A refactor might look like this:

```jsx
const formSchema = {
  fields: [
    [
      { name: 'firstName' },
      { name: 'lastName' },
    ],
    { name: 'title' },
    { name: 'company' },
    {
      control: 'Dropdown',
      name: 'gender',
      selection: true,
      options: [
        { text: 'Male', value: 'male' },
        { text: 'Female', value: 'female' },
      ],
    },
    { control: 'Checkbox', name: 'newsletter', label: 'Interested in receiving newletter' },
    { control: 'Button', type: 'submit', name: 'Submit' },
  ],
}
```

I could see an argument for allowing a simple string for the `name`.  Not sure how feel about adding more special case magic though:

```jsx
const formSchema = {
  fields: [
    ['firstName', 'lastName'],
    'title',
    'company',
    // ...
  ],
}
```